src/google/protobuf/compiler/parser_unittest.cc

```diff
@@ -1500,6 +1534,42 @@ TEST_F(ParseErrorTest, DuplicateJsonName) {
       "1:41: Already set option \"json_name\".\n");
 }
 
+TEST_F(ParseErrorTest, MsgReservedIdentifierOnlyInEditions) {
+  ExpectHasErrors(
+      "message TestMessage {\n"
+      "  reserved foo, bar;\n"
+      "}\n",
+      "1:11: Expected field name or number range.\n");
+}
+
+TEST_F(ParseErrorTest, MsgReservedIdentifierCantMixWithStrings) {
+  ExpectHasErrors(
+      "edition = \"2023\";\n"
+      "message TestMessage {\n"
+      "  reserved foo, \"bar\";\n"
+      "}\n",
+      "2:16: Expected field name identifier.\n");
+}
+
+TEST_F(ParseErrorTest, EnumReservedIdentifierOnlyInEditions) {
+  ExpectHasErrors(
+      "enum TestEnum {\n"
+      "  FOO = 0;\n"
+      "  reserved foo, bar;\n"
+      "}\n",
+      "2:11: Expected enum value or number range.\n");
+}
+
+TEST_F(ParseErrorTest, EnumReservedIdentifierCantMixWithStrings) {
+  ExpectHasErrors(
+      "edition = \"2023\";\n"
+      "enum TestEnum {\n"
+      "  FOO = 0;\n"
+      "  reserved foo, \"bar\";\n"
+      "}\n",
+      "3:16: Expected enum value identifier.\n");
+}
+
```

Comment:
These "CantMixWithStrings" tests seem redundant.  They're actually just testing the same thing as the tests above, since the identifier is invalid.

Reply:
Whoops, misunderstood what this test was doing.  Can you add tests for mixing field numbers as well?  IIUC those won't work

Reply:
The error message is different: the identifier is not invalid here because these include `edition = "2023";` at the top.

Reply:
Is this error because you mixed identifiers and strings or because you used string at all?  If it's the latter, we should remove the identifier from this test.  If it's the former, we should add a test that mixes identifiers and field numbers and a dedicated test for the latter

Reply:
It's because it has a mix: since the first item was an _identifier_, it wants all subsequent items to be identifiers. It actually does this whenever there is a comma but the next token is not an identifier. (The existing functionality for numbers and strings is very similar: if the the next token is not a number of string literal, depending on context, it emits a similar message.)

I added the word "identifier" here, hoping to make it clear that an identifier is needed (in case they instead put a string, like in this example). I can also update existing messages for string literals to clarify they want a string literal.